<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro"  name="spm_3dof">

    <!-- This will include all the contents of example_include.xacro first. Go check it out! -->
    <xacro:include filename="example_include.xacro" />

    <!-- Load spm_3dof geometric parameters: -->
    <xacro:arg name="r_c" default="1.0" />
    <xacro:arg name="ang_b_m" default="1.0" />
    <xacro:arg name="r_b" default="1.0" />
    <xacro:arg name="d" default="1.0" />
    <xacro:arg name="r_e" default="1.0" />

    <xacro:arg name="r_s_piv" default="1.0" />
    <xacro:arg name="r_s_m" default="1.0" />
    <xacro:arg name="r_s_elb" default="1.0" />
    <xacro:arg name="r_s_epl" default="1.0" />

    <!-- Convert them into xacro properties: -->
    <xacro:property name="mm2m" value="0.001" />

    <xacro:property name="r_c" value="$(arg r_c)" />
    <xacro:property name="ang_b_m" value="$(arg ang_b_m)" />
    <xacro:property name="r_b" value="$(arg r_b)" />
    <xacro:property name="d" value="$(arg d)" />
    <xacro:property name="r_e" value="$(arg r_e)" />

    <xacro:property name="r_s_piv" value="$(arg r_s_piv)" />
    <xacro:property name="r_s_m" value="$(arg r_s_m)" />
    <xacro:property name="r_s_elb" value="$(arg r_s_elb)" />
    <xacro:property name="r_s_epl" value="$(arg r_s_epl)" />
    

    <!-- world -->
    <link name="world"></link>


    <!-- base -->
    <link name="spm_base">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <mesh filename="package://cga_ik/meshes/spm_3dof/base_plate.stl" 
                      scale="${r_b} ${r_b} ${r_b}"/>
            </geometry>
            <material name="white" />
        </visual>
    </link>

    <!-- world to base -->
    <joint name="joint_world_base" type="fixed">
        <parent link="world"/>
        <child link="spm_base"/>   
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>


    <!-- rot_cen -->
    <link name="spm_rot_cen"></link>

    <!-- base to rot_cen -->
    <joint name="joint_base_rot_cen" type="floating">
        <parent link="spm_base"/>
        <child link="spm_rot_cen"/>   
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>


    <!-- motor_0 -->
    <link name="spm_motor_0">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 ${pi}"/>
            <geometry>
                <mesh filename="package://cga_ik/meshes/spm_3dof/arc_link_thin.stl" 
                    scale="${r_s_m} 
                           ${r_s_m} 
                           ${r_s_m}"/>
            </geometry>
            <material name="white" />
        </visual>
    </link>

    <!-- rot_cen to motor_0 -->
    <joint name="joint_rot_cen_motor_0" type="floating">
        <parent link="spm_rot_cen"/>
        <child link="spm_motor_0"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>


    <!-- motor_1 -->
    <link name="spm_motor_1">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 ${pi}"/>
            <geometry>
                <mesh filename="package://cga_ik/meshes/spm_3dof/arc_link_thin.stl" 
                    scale="${r_s_m} 
                           ${r_s_m} 
                           ${r_s_m}"/>
            </geometry>
            <material name="white" />
        </visual>
    </link>

    <!-- rot_cen to motor_1 -->
    <joint name="joint_rot_cen_motor_1" type="floating">
        <parent link="spm_rot_cen"/>
        <child link="spm_motor_1"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>


    <!-- motor_2 -->
    <link name="spm_motor_2">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 ${pi}"/>
            <geometry>
                <mesh filename="package://cga_ik/meshes/spm_3dof/arc_link_thin.stl" 
                    scale="${r_s_m} 
                           ${r_s_m} 
                           ${r_s_m}"/>
            </geometry>
            <material name="white" />
        </visual>
    </link>

    <!-- rot_cen to motor_2 -->
    <joint name="joint_rot_cen_motor_2" type="floating">
        <parent link="spm_rot_cen"/>
        <child link="spm_motor_2"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>


    <!-- elbow_0 -->
    <link name="spm_elbow_0">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <mesh filename="package://cga_ik/meshes/spm_3dof/arc_link_thin.stl" 
                    scale="${r_s_elb} 
                           ${r_s_elb} 
                           ${r_s_elb}"/>
            </geometry>
            <material name="white" />
        </visual>
    </link>

    <!-- rot_cen to elbow_0 -->
    <joint name="joint_rot_cen_elbow_0" type="floating">
        <parent link="spm_rot_cen"/>
        <child link="spm_elbow_0"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>


    <!-- elbow_1 -->
    <link name="spm_elbow_1">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <mesh filename="package://cga_ik/meshes/spm_3dof/arc_link_thin.stl" 
                    scale="${r_s_elb} 
                           ${r_s_elb} 
                           ${r_s_elb}"/>
            </geometry>
            <material name="white" />
        </visual>
    </link>

    <!-- rot_cen to elbow_1 -->
    <joint name="joint_rot_cen_elbow_1" type="floating">
        <parent link="spm_rot_cen"/>
        <child link="spm_elbow_1"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>


    <!-- elbow_2 -->
    <link name="spm_elbow_2">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <mesh filename="package://cga_ik/meshes/spm_3dof/arc_link_thin.stl" 
                    scale="${r_s_elb} 
                           ${r_s_elb} 
                           ${r_s_elb}"/>
            </geometry>
            <material name="white" />
        </visual>
    </link>

    <!-- rot_cen to elbow_2 -->
    <joint name="joint_rot_cen_elbow_2" type="floating">
        <parent link="spm_rot_cen"/>
        <child link="spm_elbow_2"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>


    <!-- epl_c -->
    <link name="spm_epl_c">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <mesh filename="package://cga_ik/meshes/spm_3dof/end_plate.stl" 
                      scale="${r_e} ${r_e} ${r_e}"/>
            </geometry>
            <material name="white" />
        </visual>
    </link>

    <!-- rot_cen to epl_c -->
    <joint name="joint_rot_cen_epl_c" type="floating">
        <parent link="spm_rot_cen"/>
        <child link="spm_epl_c"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>
    

</robot>
